packages/react/src/time-input/TimeInput.tsx

```diff
@@ -0,0 +1,117 @@
+import {
+	forwardRef,
+	type ChangeEvent,
+	type FocusEvent,
+	useEffect,
+	useMemo,
+	useState,
+} from 'react';
+import { Field } from '../field';
+import { TextInputProps, textInputStyles } from '../text-input';
+import {
+	acceptedTimeFormats,
+	formatTime,
+	transformValuePropToInputValue,
+	type TimeFormat,
+} from './utils';
+
+// This const object is required for the tests and also makes the component slightly more performant
+const STABLE_DEFAUT_VALUE = {
+	formatted: '',
+	raw: '',
+};
+
+export const TimeInput = forwardRef<HTMLInputElement, TimeInputProps>(
+	function TimeInput(
+		{
+			block,
+			hideOptionalLabel,
+			hint,
+			id,
+			invalid,
+			label,
+			maxWidth = 'md',
+			message,
+			onBlur: onBlurProp,
+			onChange: onChangeProp,
+			required,
+			timeFormat: timeFormatProp = 'h:mm aaa',
+			value = STABLE_DEFAUT_VALUE,
+			...props
+		},
+		ref
+	) {
+		const [inputValue, setInputValue] = useState(
+			transformValuePropToInputValue(value?.raw, timeFormatProp)
+		);
+		// Update the input when the value and/or timeFormat updates
+		useEffect(() => {
+			setInputValue(transformValuePropToInputValue(value.raw, timeFormatProp));
+		}, [timeFormatProp, value]);
+
+		const onBlur = (e: FocusEvent<HTMLInputElement>) => {
+			const inputValue = e.target?.value;
+			const normalizedTime = formatTime(inputValue, 'HH:mm');
+
+			onChangeProp?.({
+				formatted: formatTime(inputValue, timeFormatProp),
+				raw: normalizedTime,
+			});
+
+			onBlurProp?.(e);
+		};
+
+		const onChange = (e: ChangeEvent<HTMLInputElement>) => {
+			const inputValue = e.target?.value;
+			// Immediately update the input field
+			setInputValue(inputValue);
+		};
+
+		const secondaryLabel = useMemo(() => {
+			const timeFormat = acceptedTimeFormats.includes(timeFormatProp)
+				? timeFormatProp
+				: 'h:mm aaa';
+			return '(e.g. ' + formatTime('21:30', timeFormat) + ')';
+		}, [timeFormatProp]);
+
+		const styles = textInputStyles({ block, maxWidth });
+
+		return (
+			<Field
+				hideOptionalLabel={hideOptionalLabel}
+				hint={hint}
+				id={id}
+				invalid={invalid}
+				label={label}
+				maxWidth={maxWidth}
+				message={message}
+				required={Boolean(required)}
+				secondaryLabel={secondaryLabel}
+			>
+				{(a11yProps) => (
+					<input
+						{...a11yProps}
+						css={styles}
+						onBlur={onBlur}
+						onChange={onChange}
+						ref={ref}
+						type="text"
+						value={inputValue}
+						{...props}
+					/>
+				)}
+			</Field>
+		);
+	}
+);
+
+export type TimeValue = {
```

Comment:
Good idea on the instructions. I don't think iso format is useful here because 1) the native html time input doesn't do this and 2) it's just plain text. It has no real meaning and correlation to any timezone or whatever. A consuming app can easily use the 24 hour time and convert it to whatever they may need

Reply:
ok, please add the description text to `formatted` and `raw` though.